### ISSUE TITLE
Discussion on InChI Support

### ISSUE THREAD

**Issue by frodofine** [chemfiles/chemfiles#188]
In the next few days, I am going to submit a PR to introduce [InChI](https://www.inchi-trust.org/) support as a new `Format` class (read and write support currently added). All the code for generating and reading **InChI** strings is contained within this class and therefore not accessible to other methods. This is unfortunate as **InChI** support could be useful to export through the CAPI and used in other projects that require standardization of chemical topology. There are some caveats to this I wish to discuss.

First, an **InChI** string can be constructed with or without chirality support. If one decides to incorporate Chirality, they may do so via explicitly giving the directionality of bonds (e.g. UP, DOWN, E/Z), or by supplying the 3D location of all atoms in the molecule.  The first case is explained in the API as Stereo0D. In the second case, the **InChI** library will assign the chirality using 3D calculations. Therefore a public interface for *chemfiles* should offer two methods for the generation of **InChI** strings, one that accepts a `Topology` argument only (no 3D information) and constructs the string using the stereo3D interface. A second method will accept a `Frame` argument and generates the string with the 3D coordinates. 

Second, auxiliary data can be supplied using an additional **AuxInfo** string. This string can hold 3D coordinate data and other useful characteristics of a given molecule. A simple fix is to simply have **chemfiles** return multiple strings instead of one. Same goes for reading the given strings.

Third, the **InChI** string generation can be performed using a variety of options. Therefore we will need to incorporate a way to provide an options string for **InChI** generation. This could be handled by an additional argument to API or as options in a configuration file.

Finally, **InChI** keys can be generated from the **InChI** strings. I propose we simply add a function to hash an already generated **InChI** string as is done in OpenBabel and RDKit.

**Comment by Luthaf** [comment]
So, I am not really familiar with InChI: I've heard of it, but never used it, or even seen an InChI string ... I definitively think that having InChI support is great, especially as it seems to be a well used format. 

However, some of your points/questions raises another question for me:

> This is unfortunate as InChI support could be useful to export through the CAPI and used in other projects that require standardization of chemical topology.

> Therefore we will need to incorporate a way to provide an options string for InChI generation. 

> I propose we simply add a function to hash an already generated InChI string as is done in OpenBabel and RDKit.

Why should this be done inside chemfiles ? If people need to use InChI from C, or use advanced options, why can't they use the InChI libraries ?

My stand on configurable formats is to read everything possible, but pick a single (the best one possible) way when writing files. For example, there are two variants of the `.arc` Tinker format, one with unit cell and one without. We read both, but always use the version with unit cell when writing. 

So for writing InChI, could we use the AuxInfo string to store coordinates, and pick a sensible default for the string generation options.

What would be your use cases for make InChI special here ?

**Comment by frodofine** [comment]
InChI is a great format for standardizing topologies. It was created by IUPAC and is used to look up compounds in a wide variety of online databases (what InChI keys are designed to do).

> If people need to use InChI from C, or use advanced options, why can't they use the InChI libraries?

What I want is the ability to use a function like this:

`int get_inchi(const CHFL_FRAME*, char*, size_t* );`
`int get_inchi_key(const CHFL_FRAME*, char*, size_t*);`

This way I can easily get the InChI key of a molecule I'm working on, in say Python, and query it in PubChem using that key. My current implementation requires me to write out a file and read it back in - gross.

> My stand on configurable formats is to read everything possible, but pick a single (the best one possible) way when writing files.

That's fine. In fact, using options is considered 'non-standard' and generates a string that cannot be used to make an InChI key. Therefore, using no options is best for most cases.

> What would be your use cases for make InChI special here?

Database searching by the InChI key.

**Comment by Luthaf** [comment]
Ok, thanks for elaborating ! What do you think of a solution using in memory files, like #187 ? This would solve the same problem while being more generic.

The code would be only a bit longer, something like
```cpp
char buffer[2048];
CHFL_TRAJECTORY* traj = chfl_trajectory_buffer("InChI", 'w', buffer, 2048);
chfl_trajectory_write(traj, frame);
```

You could then easily build the `get_inchi`, `get_inchi_key` function in your own code from this abstraction.

What do you think?

**Comment by frodofine** [comment]
The only issue with this solution is that some compounds do not have 3D coordinates, therefore I wanted to implement a second method like the following (I know C doesn't support overloading) :

```C
int get_inchi(const CHFL_TOPOLOGY*, char*, size_t* );
int get_inchi_key(const CHFL_TOPOLOGY*, char*, size_t*);
```

These functions would use Atom/Bond properties to assign chirality roles instead of the 3D coordinates.  Similarly, there could be:

```C
int read_inchi(CHFL_TOPOLOGY*, char* inchi);
int read_inchi(CHFL_FRAME*, char* inchi, char* auxinfo);
```

This is meant to be a solution for Stereo0D features.


An alternative would be to set a `Frame` property like *use_stereo0d* when outputting the file, but I wanted to take advantage of Chemfile's separation of `Topology` and 3D information.

**Comment by Luthaf** [comment]
For Stereo0D, would it be possible to choose to use it if all atomic positions are set to 0 in the frame ? Then you can easily create the `int get_inchi(const CHFL_TOPOLOGY*, char*, size_t* );` overload, simply creating an empty frame of the right size and setting its topology.

Also, how would you give the bond orientation information for Stereo0D?

**Comment by frodofine** [comment]
Yes, that would work, but it does require one to check all the positions for zero (not the worst thing in the world).

There's an `INCHI_StereoType_DoubleBond` and `INCHI_BOND_STEREO_SINGLE_1DOWN` flags one can set.

**Comment by Luthaf** [comment]
But you need to set the flags on bonds, right ? This would need to be stored somewhere, either as bond types or as bonds "properties"?

**Comment by frodofine** [comment]
Currently, I'm storing this as a different bond type.

**Comment by schneiderfelipe** [comment]
This is exactly what I'm looking for! Any progress on this so far?

**Comment by frodofine** [comment]
Unfortunately I don’t have time to complete this PR as part of my new position (I would if I did through!). I might have time next month, but to be honest I don’t code much in C++ anymore, so it might take some time to relearn some of those skills.

I’m perfectly willing to help explain my thinking and rationale behind what is missing to complete this though!

**Comment by schneiderfelipe** [comment]
Hey thank you all for the great work! Unfortunately I don't have time to do this right now as well. Sorry!
